pkg/auth/handler/loginid/add.go

```diff
@@ -0,0 +1,171 @@
+}
+
+// @JSONSchema
+const AddLoginIDRequestSchema = `
+{
+	"$id": "#AddLoginIDRequest",
+	"type": "object",
+	"properties": {
+		"key": { "type": "string", "minLength": 1 },
+		"value": { "type": "string", "minLength": 1 }
+	},
+	"required": ["key", "value"]
+}
+`
+
+/*
+	@Operation POST /login_id/add - Add login ID
+		Add new login ID for current user.
+
+		@Tag User
+		@SecurityRequirement access_key
+		@SecurityRequirement access_token
+
+		@RequestBody
+			Describe the new login ID.
+			@JSONSchema {AddLoginIDRequest}
+
+		@Response 200 {EmptyResponse}
+
+		@Callback identity_create {UserSyncEvent}
+		@Callback user_sync {UserSyncEvent}
+*/
+type AddLoginIDHandler struct {
+	Validator                *validation.Validator      `dependency:"Validator"`
+	AuthContext              coreAuth.ContextGetter     `dependency:"AuthContextGetter"`
+	RequireAuthz             handler.RequireAuthz       `dependency:"RequireAuthz"`
+	AuthInfoStore            authinfo.Store             `dependency:"AuthInfoStore"`
+	PasswordAuthProvider     password.Provider          `dependency:"PasswordAuthProvider"`
+	IdentityProvider         principal.IdentityProvider `dependency:"IdentityProvider"`
+	UserVerificationProvider userverify.Provider        `dependency:"UserVerificationProvider"`
+	TxContext                db.TxContext               `dependency:"TxContext"`
+	UserProfileStore         userprofile.Store          `dependency:"UserProfileStore"`
+	HookProvider             hook.Provider              `dependency:"HookProvider"`
+}
+
+func (h AddLoginIDHandler) ProvideAuthzPolicy() authz.Policy {
+	return policy.AllOf(
+		authz.PolicyFunc(policy.DenyNoAccessKey),
+		policy.RequireValidUser,
+	)
+}
+
+func (h AddLoginIDHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	err := h.Handle(w, r)
+	if err == nil {
+		handler.WriteResponse(w, handler.APIResponse{Result: struct{}{}})
+	} else {
+		handler.WriteResponse(w, handler.APIResponse{Error: err})
+	}
+}
+
+func (h AddLoginIDHandler) Handle(w http.ResponseWriter, r *http.Request) error {
+	var payload AddLoginIDRequestPayload
+	if err := handler.BindJSONBody(r, w, h.Validator, "#AddLoginIDRequest", &payload); err != nil {
+		return err
+	}
+
+	err := hook.WithTx(h.HookProvider, h.TxContext, func() error {
+		authInfo, _ := h.AuthContext.AuthInfo()
+		userID := authInfo.ID
+
+		principals, err := h.PasswordAuthProvider.GetPrincipalsByUserID(userID)
+		if err != nil {
+			return err
+		}
+
+		newLoginID := password.LoginID{Key: payload.Key, Value: payload.Value}
+		loginIDs := extractLoginIDs(principals)
+		newLoginIDIndex := len(loginIDs)
+		loginIDs = append(loginIDs, newLoginID)
+		err = validateLoginIDs(h.PasswordAuthProvider, loginIDs, newLoginIDIndex)
+		if err != nil {
+			return err
+		}
+
+		newPrincipal, err := h.PasswordAuthProvider.MakePrincipal(userID, "", newLoginID, password.DefaultRealm)
+		if err != nil {
+			return err
+		}
+		newPrincipal.HashedPassword = principals[0].HashedPassword
```

Comment:
You're right. I can identify two problems for users without existing password principals at the moment:
1. We cannot fill in a correct hashed password, and user cannot change password.
2. Misconfiguration to required login ID amount may cause inability to add login ID.

For problem 1, the previously discussed solution is to create a password principals without a valid password hash (i.e. user cannot login using it), and password can be changed by re-authenticating using SSO. I'll implement the first part first in this PR.

For problem 2, we have error message showing the login ID amount constraint violation (e.g. not enough 'email' login ID), so developer can understand what's going on and fix the config (i.e. reduce the minimum amount of 'email' login ID to 0). Note that this would not occur in default configuration.